### PR TITLE
Add Rubocop rule to enforce Meterpreter command requirements in modules

### PR DESCRIPTION
**Pull request by cgranleese-r7**
# WIP
Continuation from work started in #15079

This PR adds a new Rubocop rule along with tests. The cop will enforce that when ran against Metasploit modules, it will generate a Meterpreter commands list if there are any Meterpreter api calls present within said module.

The current plan is to leave this PR open and run the cop against subsections of modules and have those reviewed and landed. Then any feedback can be implemented in this PR. Once the cop has been ran against the necessary modules we can then come back and review/land this PR as a final step.

## Note
As seen in the tests, there is support to alter/add sections of methods depending on various situations. e.g. 
"The cop verifies if a there is an initialise method, if no initialise present, it should be generated and appended appropriately."

I have tried to identify as many of these scenarios as possible, but if there are anymore edge-cases that may not have been covered in the tests, please call them out and I can get tests added as needed 🕵️

### CODE REVIEW

File: lib/msf/core/post/hardware/automotive/uds.rb [hunk 1]
```diff
@@ -6,6 +6,16 @@ module Automotive
...
 
+  def initialize(info = {})
+    super(update_info(
+            info,
+            'Compat' => { 'Meterpreter' => { 'Commands' => %w{
+              automotive_active_bus
+              automotive_send_isotp_and_wait_for_response
+      } } }
+          ))
+  end
```
Comment (adfoster-r7) [inline lib/msf/core/post/hardware/automotive/uds.rb:17]: Rubocop can fix this automatically in the future, but if you want to make the diff cleaner in the future we can pre-empt that work:
```
          def initialize(info = {})
            super(
              update_info(
                info,
                'Compat' => {
                  'Meterpreter' => {
                    'Commands' => %w[
                      automotive_active_bus
                      automotive_send_isotp_and_wait_for_response
                    ]
                  }
                }
              )
            )
          end
```

File: lib/msf/core/post/hardware/automotive/uds.rb [hunk 2]
```diff
@@ -6,6 +6,22 @@ module Automotive
...
+          'Meterpreter' => {
+            'Commands' => %w[
+              automotive_active_bus
+              automotive_send_isotp_and_wait_for_response
+            ]
+          }
+        }
+      )
+    )
+  end
```
Comment (cgranleese-r7) [inline lib/msf/core/post/hardware/automotive/uds.rb:23]: After a conversation with @zeroSteiner. I'm going to be reverting any `/hardware/...` modules that were altered 
We don't believe the `/hardware/...` modules need the Meterpreter compatibility because they don't use Meterpreter sessions.

File: lib/msf/core/post/windows/registry.rb [hunk 3]
```diff
@@ -49,6 +49,33 @@ module Registry
...
+              stdapi_sys_registry_unload_key
+              stdapi_sys_registry_create_key
+              stdapi_sys_registry_open_key
+              stdapi_sys_registry_delete_key
+              stdapi_sys_registry_enum_key_direct
+              stdapi_sys_registry_enum_value_direct
+              stdapi_sys_registry_query_value_direct
+              stdapi_sys_registry_set_value_direct
+              stdapi_sys_registry_type2str
+              stdapi_sys_registry_check_key_exists
```
Comment (smcintyre-r7) [inline lib/msf/core/post/windows/registry.rb:71]: The registry commands use the `stdapi_registry` prefix, without the `sys` keyword.

File: lib/rubocop/cop/lint/meterpreter_commands_dependencies.rb [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+        def_node_matcher :command_list?, <<~PATTERN
+          (lvasgn :commands
+            _)
+        PATTERN
+
+        extend AutoCorrector
+
+        def on_new_investigation
+          super
+          @current_commands = [] # TODO: Not sure if I need to keep track of current commands, as I can just add all to latest and rewrite the list
```
Comment (adfoster-r7) [inline lib/rubocop/cop/lint/meterpreter_commands_dependencies.rb:32]: You'll need to keep track of the existing commands, so you can add and offense when there's additional commands present that aren't required

File: modules/exploits/android/local/binder_uaf.rb [hunk 5]
```diff
@@ -42,6 +42,13 @@ def initialize(info={})
         ],
         'DisclosureDate' => '2019-09-26',
         'SessionTypes'   => [ 'meterpreter' ],
+        'Compat' => {
+          'Meterpreter' => {
```
Comment (adfoster-r7) [inline modules/exploits/android/local/binder_uaf.rb:46]: TODO: Confirm if we need to generate this only if `sessiontypes` has `meterpreter` in it

File: lib/rubocop/cop/lint/meterpreter_commands_dependencies.rb [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+                "\n#{commands_whitespace}'Commands' => %w[" \
+                "\n#{array_content_whitespace}#{@identified_commands.join("\n#{array_content_whitespace}")}" \
+                "\n#{commands_whitespace}]" \
+                "\n#{meterpreter_whitespace}}" \
+                "\n#{info_whitespace}}" \
+                "\n#{update_info_whitespace})" \
+                "\n#{super_whitespace})" \
+                "\n#{def_whitespace}end" \
+                "\n  "
+
```
Comment (adfoster-r7) [inline lib/rubocop/cop/lint/meterpreter_commands_dependencies.rb:244]: We'll want to DRY this all out, either via more clever string manipulation, or a potential alternative algorithm might be:

1. Generate the template which would have everything
```ruby
# (Formatted correctly)
def generate_initialize_template
    <<~EOF
       def initialize(...)
           super(
               update_info(
                  info,
                   {
                    compat => meterpreter => commands [
                      ...
                     ]
                 }
             )
           )
       end
    EOF
end
```
1. Parse the template again with rubocop:
```
source = RuboCop::ProcessedSource.new(code, RUBY_VERSION.to_f)
node = source.ast
```
1. Decide which nodes are present in the original source, then extract the subset of the 'everything' template that you need, and append it to the right place in the original source. Just using normal rubocop node matchers to extract the data out. Would need to ensure it's indented to the right level, but that should be pretty trivial

File: lib/rubocop/cop/lint/meterpreter_commands_dependencies.rb [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+                "#{commands_whitespace}]\n" \
+                "#{meterpreter_whitespace}}\n" \
+                "#{compat_whitespace}}"
+
+              corrector.insert_after(@end_of_info_node, test_new_hash)
+
+            elsif @compat_node.nil? && @meterpreter_node.nil? && @command_node.nil? && @initialize_node.nil?
+              # White spacing handling based of node offsets
+              def_whitespace = offset(@node_after_initialize)
+              super_whitespace = def_whitespace + "  "
```
Comment (adfoster-r7) [inline lib/rubocop/cop/lint/meterpreter_commands_dependencies.rb:220]: This should be using `indent(...)`

I think things weren't working as expected, as it reads the required whitespace from its configuration:
```ruby
# gems/rubocop-1.16.0/lib/rubocop/cop/mixin/alignment.rb
def indentation(node)
    offset(node) + (SPACE * configured_indentation_width)
end
```

You'll want to update the tests to have the configured indentation width to work:
```rspec
  let(:config) do
    RuboCop::Config.new(
      'Layout/IndentationWidth' => {
        'Width' => indentation_width
      })
  end
  let(:indentation_width) { 2 }
```

**Comment by jmartin-r7** [comment]
Let's also ensure we update the optionally loaded test modules.

File: lib/rubocop/cop/lint/meterpreter_commands_dependencies.rb [hunk 8]
```diff
@@ -0,0 +1,1220 @@
...
+
+            break
+          end
+        end
+
+        def autocorrector
+          lambda do |corrector|
+            # Removes the railgun_api call if we are already calling railgun in its entirety.
+            if @current_frame.identified_commands.include?("stdapi_railgun_*") && @current_frame.identified_commands.include?("stdapi_railgun_api*")
+              require "pry"; binding.pry
```
Comment (smcintyre-r7) [inline lib/rubocop/cop/lint/meterpreter_commands_dependencies.rb:1061]: This looks left over from debugging.
